### PR TITLE
(feat): Animal/ped shooting gun

### PR DESCRIPTION
**Pull request by maurojrr**
Pretty dumb feature, I've copied and pasted a lot of vehicle gun code and modified it to shoot peds. 

I'm a newbie and my main issue was figuring out how to create a selectable ped model list like we have in Spawn Ped feature, but couldn't  implement it (couldn't copy and paste hehe). 

Here's a video of it working
[https://youtu.be/GbtSRLXKr04](url)

Anyways, just a simple idea and suggestion, hope you guys can help and improve. Sorry for bad English.

### CODE REVIEW

**Comment by gir489returns** [comment]
While this seems trivial, this doesn't seem to be that heavy of a lift, and would require minimal overhead for us to maintain it.

Why did you add prop_thindesertfiller_aa to the list of crash_props?

File: src/core/settings.hpp [hunk 1]
```diff
@@ -5,6 +5,8 @@
 #include "core/data/ptfx_effects.hpp"
 #include "enums.hpp"
 #include "file_manager.hpp"
+#include <list>
```
Comment (gir489returns) [inline src/core/settings.hpp:8]: Pointless include.

File: src/views/self/view_weapons.cpp [hunk 2]
```diff
@@ -172,6 +173,16 @@ namespace big
 			ImGui::Checkbox("RAINBOW_PAINT"_T.data(), &g.weapons.paintgun.rainbow);
 			ImGui::SliderFloat("VIEW_WEAPON_RAINBOW_SPEED"_T.data(), &g.weapons.paintgun.speed, 0.f, 10.f);
 			if (!g.weapons.paintgun.rainbow) { ImGui::ColorEdit4("VIEW_WEAPON_PAINT_GUN_COLOR"_T.data(), g.weapons.paintgun.col); }
+			break;
+		case CustomWeapon::SHARK_GUN:
+			static char animal_gun[32];
+			std::memcpy(animal_gun, g.weapons.animal.c_str(), 32);
+			if (ImGui::InputTextWithHint("SHOOTING_MODEL"_T.data(), "NAME_ANIMAL_MODEL"_T.data(), animal_gun, sizeof(animal_gun)))
```
Comment (gir489returns) [inline src/views/self/view_weapons.cpp:180]: Use components::input_text_with_hint instead of invoking the ImGui item directly. Also use IM_ARRAYSIZE on animal_gun as it is a hard requirement from ImGui.

File: src/views/self/view_weapons.cpp [hunk 3]
```diff
@@ -172,6 +173,16 @@ namespace big
...
+			{
+				g.weapons.animal = animal_gun;
+			}
+			if (ImGui::IsItemActive())
```
Comment (gir489returns) [inline src/views/self/view_weapons.cpp:184]: components::input_text_with_hint will handle this for you.

File: src/util/protection.cpp [hunk 4]
```diff
@@ -4,7 +4,7 @@
...
...("proc_forest_grass01"), RAGE_JOAAT("prop_small_bushyba"), RAGE_JOAAT("v_res_d_dildo_a"), RAGE_JOAAT("v_res_d_dildo_b"), RAGE_JOAAT("v_res_d_dildo_c"), RAGE_JOAAT("v_res_d_dildo_d"), RAGE_JOAAT("v_res_d_dildo_e"), RAGE_JOAAT("v_res_d_dildo_f"), RAGE_JOAAT("v_res_skateboard"), RAGE_JOAAT("prop_battery_01"), RAGE_JOAAT("prop_barbell_01"), RAGE_JOAAT("prop_barbell_02"), RAGE_JOAAT("prop_bandsaw_01"), RAGE_JOAAT("prop_bbq_3"), RAGE_JOAAT("v_med_curtainsnewcloth2"), RAGE_JOAAT("bh1_07_flagpoles"), RAGE_JOAAT("hash_058a7eb5_deihiws_collision"), RAGE_JOAAT("proc_dry_plants_01"), RAGE_JOAAT("proc_leafyplant_01"), RAGE_JOAAT("proc_grassplantmix_02"), RAGE_JOAAT("proc_dryplantsgrass_01"), RAGE_JOAAT("proc_dryplantsgrass_02"), RAGE_JOAAT("proc_dryplantsgrass_02"), RAGE_JOAAT("proc_grasses01"), RAGE_JOAAT("prop_dryweed_002_a"), RAGE_JOAAT("prop_fernba"), RAGE_JOAAT("prop_weed_001_aa"), RAGE_JOAAT("urbangrnfrnds_01"), RAGE_JOAAT("urbanweeds01"), RAGE_JOAAT("prop_dandy_b"), RAGE_JOAAT("v_proc2_temp"), RAGE_JOAAT("prop_fernbb"), RAGE_JOAAT("proc_drygrassfronds01"), RAGE_JOAAT("prop_log_ae"), RAGE_JOAAT("prop_grass_da"), RAGE_JOAAT("prop_fragtest_cnst_04"), RAGE_JOAAT("prop_thindesertfiller_aa")};
```
Comment (gir489returns) [inline src/util/protection.cpp:7]: What is the purpose of this change?

File: src/util/protection.cpp [hunk 5]
```diff
@@ -85,4 +85,4 @@ namespace big::protection
 				return true;
 		return false;
 	}
-}
```
Comment (gir489returns) [inline src/util/protection.cpp:88]: Pointless whitespace change.

File: src/backend/looped/weapons/shark_gun.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+				Vector3 rot = CAM::GET_GAMEPLAY_CAM_ROT(rotation_order);
+				float pitch = math::deg_to_rad(rot.x); // vertical
+				//float roll = rot.y;
+				float yaw = math::deg_to_rad(rot.z + 90); // horizontal
+
+				float dist = 10.f;
+				location.x += dist * cos(pitch) * cos(yaw);
+				location.y += dist * sin(yaw) * cos(pitch);
+				location.z += dist * sin(pitch);
```
Comment (gir489returns) [inline src/backend/looped/weapons/shark_gun.cpp:39]: These calculations are wildly inefficient. 

Since rotation_order is a constant value (2), you can precompute the trigonometric functions for the fixed rotation order outside the loop (if applicable). This avoids redundant calculations during each iteration.

The expression cos(pitch) * cos(yaw) appears twice. You can compute it once and store it in a temporary variable to avoid redundant calculations.

Similarly, sin(yaw) * cos(pitch) can be computed once and reused. You also subsequently calling sine and cosine again and again after for velocity. You are basically putting forcing the compiler to generate awful code.

**Review by gir489returns — CHANGES_REQUESTED** [review]

**Comment by ToXicVibezz** [comment]
I use a ped gun for random ped // 			Ped playerPed = PLAYER::PLAYER_PED_ID();

			if (PED::IS_PED_SHOOTING(PLAYER::PLAYER_PED_ID()))
			{
				Vector3 coords;
				if (WEAPON::GET_PED_LAST_WEAPON_IMPACT_COORD(PLAYER::PLAYER_PED_ID(), &coords))
				{
					PED::CREATE_RANDOM_PED(coords.x, coords.y, coords.z);
					bPedGun = b;
				}
			}

File: src/util/protection.cpp (hunk 5 above)
Comment (Yimura) [reply src/util/protection.cpp:88]: It might be a result of difference in IDE setup, some IDE's add a newline at the end some remove it.

File: src/backend/looped/weapons/shark_gun.cpp (hunk 6 above)
Comment (Yimura) [reply src/backend/looped/weapons/shark_gun.cpp:39]: Probably a remnant from my awful code in the past, this can indeed be fixed by assigning the return values into temporary variables.

File: src/util/protection.cpp (hunk 4 above)
Comment (ghost) [reply src/util/protection.cpp:7]: That i believe is from the Burger king foot lettuce crash stand has.

**Comment by gir489returns** [comment]
@maurojrr Are you still interested in having this merged? If we don't hear from you within a few days, we will be closing this branch.

**Comment by ghost** [comment]
I think enough time has passed to answer. @gir489returns it's time to do it
